atc/atccmd/command.go

```diff
@@ -1808,7 +1821,7 @@ func (cmd *RunCommand) constructAPIHandler(
 			logger,
 			accessFactory,
 			tokenVerifier,
-			cacher,
+			teamsCacher,
 			batcher,
```

Comment:
These three components `tokenVerifier`, `teamsCacher` and `batcher` are all consumed by the `AccessorHandler` and they mostly just get orchestrated to pass data to the `accessFactory`. Furthermore, they all contain optimizations intended to reduce stress on the database -- the `tokenVerifier` now wraps a 1MB-LRU `claimsCacher`, the `teamsCacher` is of fixed size with a time-based retention policy, and the `batcher` stores user info in memory until a certain batch size is reached or a certain amount of time has passed. All of these optimization strategies are slightly different, but they all seem to relate to the core problem that "figuring out token stuff is expensive".

Now that we're not only verifying the signature of tokens but we're storing them in the DB, can we consolidate these components so use a more consistent optimization pattern?

My most-aesthetic idea was to add columns to the `users` table for:
* the current access token -- which includes its expiry time (a string? or something more indexable?)
* the user's team roles (a jsonb? if it's too big maybe move into a join table??)
* whether the user has the system claim -- this might be where the idea breaks down because a _token_ can have a system claim without actually being associated to a _user_.

It doesn't have to be this exact solution, but it still probably makes sense to cache the results of computing "is this token valid? if so, what are its team memberships and is it a system token?" (which is all the DB-bound portions of constructing an `Accessor` I think) all together, rather than having separate caches for tokens and teams.

Reply:
I think it still makes sense for there to be an `access_tokens` table, rather than adding the current access token to the `users` table, since a user can have multiple active access tokens (e.g. otherwise, a `fly login` could invalidate the token from the UI).

If I'm understanding "user's team roles" correctly, that sounds like it'd be going against (one of the) initial motivations for the new auth introduced in 6.1 - team config changes shouldn't require logging in again. I think it still makes sense to store the claims from dex in some raw form in the DB (currently, it's a `jsonb` in the `access_tokens` table), but we could cache team memberships and whether it's a system token rather than just caching the claims themselves.

Also, it's worth noting we can probably get rid of the batcher optimization now (as well as updating `users.last_login` on every request), and instead use the `iat` claim of the user's latest access token to determine this information. This is a slight change in behaviour, but it's more accurate to the name `last_login`, anyway

Reply:
Yup all good points. The cache doesn't have to be an exact duplicate of the db table.

Reply:
I pushed a commit that removes the `batcher`, and updates the `last_login` when a user actually logs in (rather than on every request). Initially I suggested this could be done using the "`iat` claim on the user's latest access token" but realized this wouldn't work since we GC expired access tokens.

It makes sense to have the `accessFactory` consume the `tokenVerifier` and `teamsCacher` (instead of the `AccessorHandler`). I'm not exactly sure how best to go about consolidating those two caches into a precomputed team membership + metadata cache. It feels like doing so would just make both optimizations less effective - whenever we need to compute a new user's memberships, we need to fetch a list of all teams; and whenever we fetch a list of teams (or invalidate the list based on time or a DB notification), we'd have to invalidate the full cache.

So, unless you're opposed, I think I'll keep the two caches separate for now